File: feature/dynamic binding/DynamicBinding.cpp
#include <iostream>

using namespace std;

class Binding {
    public:
    void binding() {
        cout << "This is Class: Binding" << endl;
    }
    virtual void dynamicBinding() {
        cout << "This is Class: Binding" << endl;
    }
};

class DynamicBinding : public Binding {
    public:
    void binding() {
        cout << "This is Class: DynamicBinding" << endl;
    }

    virtual void dynamicBinding() {
        cout << "This is Class: DynamicBinding" << endl;
    }
};

class SecondDynamicBinding : public DynamicBinding {
    public:
    void binding() {
        cout << "This is Class: SecondDynamicBinding" << endl;
    }
    
    virtual void dynamicBinding() {
        cout << "This is Class: SecondDynamicBinding" << endl;
    }
};

int main() {
    Binding* pBinding = new Binding();
    DynamicBinding* pDynamicBinding = new DynamicBinding();
    SecondDynamicBinding* pSecondDynamicBinding = new SecondDynamicBinding();
    Binding* pTest1 = pDynamicBinding;
    Binding* pTest2 = pSecondDynamicBinding;

    cout << "The first part" << endl;
    pBinding->binding();
    pDynamicBinding->binding();
    pTest1->binding();
    pTest2->binding();

    cout << "The second part" << endl;
    pBinding->dynamicBinding();
    pDynamicBinding->dynamicBinding();
    pTest1->dynamicBinding();
    pTest2->dynamicBinding();

}

File: feature/dynamic binding/DynamicBinding.java
class Binding {
    protected String test = "Binding test";

    public String getStr() {
        return test;
    }

}

class DynamicBinding extends Binding {
    protected String test = "DynamicBinding test";

    public String getStr() {
        return test;
    }
}

class SecondDynamicBinding extends DynamicBinding {
    protected String test = "SecondDynamicBinding test";

    public String getStr() {
        return test;
    }

    public static void main(String[] args) {
        Binding test1 = new Binding();
        Binding test2 = new DynamicBinding();
        Binding test3 = new SecondDynamicBinding();

        System.out.println("The first part");
        System.out.println("The class is " + test1.test);
        System.out.println("The class is " + test2.test);
        System.out.println("The class is " + test3.test);

        System.out.println("The second part");
        System.out.println("The class is " + test1.getStr());
        System.out.println("The class is " + test2.getStr());
        System.out.println("The class is " + test3.getStr());
    }
}